Sources/Yams/Representer.swift

```diff
@@ -99,31 +99,36 @@ extension Date: ScalarRepresentable {
     }
 
     private var iso8601String: String {
-#if !_runtime(_ObjC) && !swift(>=5.0)
-        // swift-corelibs-foundation has bug with nanosecond.
-        // https://bugs.swift.org/browse/SR-3158
-        return iso8601Formatter.string(from: self)
-#else
-        let calendar = Calendar(identifier: .gregorian)
-        let nanosecond = calendar.component(.nanosecond, from: self)
-        if nanosecond != 0 {
-            return iso8601WithFractionalSecondFormatter.string(from: self)
-                .trimmingCharacters(in: characterSetZero) + "Z"
-        } else {
-            return iso8601Formatter.string(from: self)
-        }
-#endif
+        let (integral, millisecond) = timeIntervalSinceReferenceDate.separateFractionalSecond(withPrecision: 3)
+        guard millisecond != 0 else { return iso8601Formatter.string(from: self) }
+
+        let dateWithoutMillisecond = Date(timeIntervalSinceReferenceDate: integral)
+        return iso8601WithoutZFormatter.string(from: dateWithoutMillisecond) +
+            String(format: ".%03d", millisecond).trimmingCharacters(in: characterSetZero) + "Z"
     }
 
     private var iso8601StringWithFullNanosecond: String {
-        let calendar = Calendar(identifier: .gregorian)
-        let nanosecond = calendar.component(.nanosecond, from: self)
-        if nanosecond != 0 {
-            return iso8601WithoutZFormatter.string(from: self) +
-                String(format: ".%09d", nanosecond).trimmingCharacters(in: characterSetZero) + "Z"
-        } else {
-            return iso8601Formatter.string(from: self)
-        }
+        let (integral, nanosecond) = timeIntervalSinceReferenceDate.separateFractionalSecond(withPrecision: 9)
+        guard nanosecond != 0 else { return iso8601Formatter.string(from: self) }
+
+        let dateWithoutNanosecond = Date(timeIntervalSinceReferenceDate: integral)
+        return iso8601WithoutZFormatter.string(from: dateWithoutNanosecond) +
+            String(format: ".%09d", nanosecond).trimmingCharacters(in: characterSetZero) + "Z"
+    }
+}
+
+private extension TimeInterval {
+    /// Separates the time interval into integral and fractional components, then rounds the `fractional` component to `precision` number of digits.
+    /// - returns: Tuple of integral part and converted fractional part
+    func separateFractionalSecond(withPrecision precision: Int) -> (integral: TimeInterval, fractional: Int) {
+        var integral = 0.0
+        let fractional = modf(self, &integral)
+        let radix = pow(10.0, Double(precision))
+        let rounded = Int((fractional * radix).rounded())
```

Comment:
P.S

It should be truncated (`.rounded(.down)`?) instead of `rounded()`.

[The Data Formatting Guide](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/DataFormatting/Articles/dfDateFormatting10_4.html#//apple_ref/doc/uid/TP40002369-SW13) said:
> The format string uses the format patterns from the Unicode Technical Standard #35. 

And [the Unicode Technical Standard #35](http://www.unicode.org/reports/tr35/tr35-31/tr35-dates.html#Date_Field_Symbol_Table) said:

|Field|Sym.|No.|Example|Description|
|-----|----|---|-------|-----------|
|second|S|1..n|3456|Fractional Second - **truncates** (like other time fields) to the count of letters. (example shows display using pattern SSSS for seconds value 12.34567)|

`DateFormatter` rounds sub-millisecond and also caused # 192.
I filed this behavior and document inconsistency as a bug: https://bugs.swift.org/browse/SR-11049

For now, this implementation rounds sub-millisecond the same as `DateFormatter`, but this may have to be fixed.